### ISSUE TITLE
Bug: no-unnecessary-condition: false positives on subtype comparisons and Record ??=

### ISSUE THREAD

**Issue by chenglou** [oxc-project/tsgolint#695]
```ts
// False positives in tsgolint's no-unnecessary-condition that eslint's same rule doesn't flag.
// All of these pass with @typescript-eslint/eslint-plugin's no-unnecessary-condition.

// Bug 1: EventTarget | null vs Window (subtype relationship not recognized)
// Window extends EventTarget, so `e.target !== window` is a valid comparison.
function repro1(e: FocusEvent) {
  if (e.target !== window) return
}

// Bug 2: `string & {}` vs `string` (intersection with empty object not recognized as string)
// `string & {}` is a common pattern for string literal autocompletion while accepting any string.
function repro2(flag: string & {}, arg: string) {
  if (arg === flag) return true
  return false
}

// Bug 3: Symbol type compared after Array.isArray guard
// After `Array.isArray` doesn't narrow away the symbol branch, comparing to the symbol should be valid.
const flagPresent = Symbol()
type FlagPresent = typeof flagPresent
function repro3(queryFlag: string[] | FlagPresent) {
  if (Array.isArray(queryFlag) && queryFlag.length === 0) return true
  if (queryFlag === flagPresent) return true
  return false
}

// Bug 4: Template literal type vs string (subtype relationship not recognized)
// `${string}_6_main` is a subtype of string, so comparing them should be valid.
type PersonalizationId = `${string}_6_main` | `${string}_7_main`
function repro4(a: PersonalizationId, b: string) {
  return a === b
}

// Bug 5: ??= on Record with dynamic keys
// Record<string, T> index access can be undefined at runtime. ??= is the intended way to initialize.
function repro5() {
  const obj: Record<string, { name: string }> = {}
  const key: string = 'foo'
  obj[key] ??= { name: key }
}
```

`bun oxlint --type-aware test-no-unnecessary-condition.ts` shows the above 5 false positives whereas `bun eslint test-no-unnecessary-condition.ts` doesn't

**Comment by camc314** [comment]
0.14.1 is releasing now

**Comment by chenglou** [comment]
Sorry @camc314 found another one at #699 🫡

**Comment by camc314** [comment]
nope - don't apologies, thanks for finding and reporting these!

**Comment by osdiab** [comment]
I think I found the same thing but I am on 0.16.0. the key is a branded type from Zod.

<img width="407" height="72" alt="Image" src="https://github.com/user-attachments/assets/e353f5d4-fddf-47e3-a57b-65a5642c8ae4" />

**Comment by camc314** [comment]
@osdiab can you make a new issue please with a minimal repro?
